view: fix use-after-move when handling view update failures

psarna opened the pull request:
The code was susceptible to use-after-move if both local
and remote updates were going to be sent.
The whole routine for sending view updates is now rewritten
to avoid use-after-move.

Fixes #8830
Tests: unit(release),
       dtest(secondary_indexes_test.py:TestSecondaryIndexes.test_remove_node_during_index_build)

psarna commented:
I didn't mark this issue as fixing #8830, because the test passes locally for me, so I can't really use it to determine if the issue is gone. It's very likely the case, because @bhalevy pinpointed the problem to be a use-after-move and the previous code did indeed use a lambda after it was moved-from.

psarna commented:
For reviewers - what this patch essentially does is inlining the lambda into its only two call sites and adding a proper handler for semaphore units, so that they're only used once. The `maybe_account_failure` lambda started off as a small helper, but over time it got really bloated, and in my opinion it got more complicated than it's worth, so it's better to have two separate code paths that look quite similar, but are actually distinct.

bhalevy commented on db/view/view.cc:
```diff
@@ -1293,10 +1281,21 @@ future<> mutate_MV(
-                     maybe_account_failure = std::move(maybe_account_failure)] (future<>&& f) mutable {
-                return maybe_account_failure(std::move(f), std::move(*target_endpoint), false, updates_pushed_remote);
+                    [s = mut.s, &stats, &cf_stats, tr_state, base_token, view_token, target_endpoint, updates_pushed_remote,
+                            units = sem_units.split(sem_units.count())] (future<>&& f) mutable {
```
won't that throw "Cannot take more units than those protected by the semaphore" if both local and remote updates are needed? (like was the case triggering the use-after-move)

bhalevy replied:
Otherwise looks good to me.

psarna replied:
No - the first split that happens would leave 0 units in sem_units, making the second split a no-op. Technically using std::move would do the same, because a moved-from instance of semaphore_units is defined as having 0 units, but I consider use-after-move an antipattern, so I coded it explicitly instead.

bhalevy replied:
Right. Missed the fact fact that calling `sem_units.count()` the second time will indeed return zero.

LGTM

nyh commented:
I'll review it now, but a question on the commit message - why does it Refs #8830, not Fixes? Would be better if it "Fixes" something, to remember to consider backport.

psarna commented:
@nyh https://github.com/scylladb/scylla/pull/8834#issuecomment-857668611 :)

psarna commented:
I'm fine with changing it to "Fixes" though if it makes the backporting machinery easier

psarna commented:
Done - if it turns out not to fix the issue after all, it can be reopened manually

nyh commented on db/view/view.cc:
```diff
@@ -1266,11 +1245,20 @@ future<> mutate_MV(
-                     maybe_account_failure = std::move(maybe_account_failure),
-                     mut_ptr = std::move(mut_ptr)] (future<>&& f) {
+                    [s = mut.s, &stats, &cf_stats, tr_state, base_token, view_token, my_address, mut_ptr = std::move(mut_ptr),
+                            units = sem_units.split(sem_units.count())] (future<>&& f) {
```
I see you explained this later. It's still really wierd, but probably not wrong.

avikivity commented:
This patch has angered materialized_views_test.py:TestMaterializedViews.interrupt_build_process_with_resharding_half_to_max_test (code dump), so I'm dequeueing it.

psarna commented:
Super sneaky... a helper function took a mutation_with_schema by reference, but then moved from it inside, leaving its schema as a nullptr. It used to work before just because the original lambda managed to copy the schema before it got moved. I'll prepare v2 with fixes.

psarna commented:
At least the invariant of this series that the number of use-after-move operations should remain identical was kept.

psarna commented:
v2:
 - fixed another use-after-move introduced in this series - this time it was about a schema pointer inside `frozen_mutation_and_schema`
 - changed another fishy pass-by-reference in the same function to by-value

psarna commented:
Tests: unit(release), dtest(materialized_views_test.py:TestMaterializedViews.interrupt_build_process_with_resharding_half_to_max_test) - the one which regressed in dtest-gating
